C++ helpers depend on xtensor-blas

I've got linking errors like
```
/ld: helpers/CMakeFiles/petsird_analysis.dir/petsird_analysis.cpp.o: in function `auto xt::linalg::dot<xt::xarray_container<xt::uvector<float, std::allocator<float> >, (xt::layout_type)1, xt::svector<unsigned long, 4ul, std::allocator<unsigned long>, true>, xt::xtensor_expression_tag>, xt::xarray_container<xt::uvector<float, std::allocator<float> >, (xt::layout_type)1, xt::svector<unsigned long, 4ul, std::allocator<unsigned long>, true>, xt::xtensor_expression_tag> >(xt::xexpression<xt::xarray_container<xt::uvector<float, std::allocator<float> >, (xt::layout_type)1, xt::svector<unsigned long, 4ul, std::allocator<unsigned long>, true>, xt::xtensor_expression_tag> > const&, xt::xexpression<xt::xarray_container<xt::uvector<float, std::allocator<float> >, (xt::layout_type)1, xt::svector<unsigned long, 4ul, std::allocator<unsigned long>, true>, xt::xtensor_expression_tag> > const&)':
petsird_analysis.cpp:(.text._ZN2xt6linalg3dotINS_16xarray_containerINS_7uvectorIfSaIfEEELNS_11layout_typeE1ENS_7svectorImLm4ESaImELb1EEENS_22xtensor_expression_tagEEESB_EEDaRKNS_11xexpressionIT_EERKNSC_IT0_EE[_ZN2xt6linalg3dotINS_16xarray_containerINS_7uvectorIfSaIfEEELNS_11layout_typeE1ENS_7svectorImLm4ESaImELb1EEENS_22xtensor_expression_tagEEESB_EEDaRKNS_11xexpressionIT_EERKNSC_IT0_EE]+0x294): undefined reference to `cblas_sgemm'
```

Indeed, https://github.com/ETSInitiative/PETSIRD/blob/76d1a9196b68ac5d7178e4e2651ee96f597d78eb/cpp/helpers/include/petsird_helpers/geometry.h#L12
means we need to depend on `xtensor-blas`

--- a/cpp/CMakeLists.txt
+++ b/cpp/CMakeLists.txt
@@ -15,5 +15,5 @@ if(CCACHE_PROGRAM)
   message(STATUS "ccache found, so we will use it.")
 endif()
 
-add_subdirectory(generated)
+add_subdirectory(generated/petsird)
 add_subdirectory(helpers)

--- a/cpp/CMakeLists.txt
+++ b/cpp/CMakeLists.txt
@@ -15,5 +15,14 @@ if(CCACHE_PROGRAM)
   message(STATUS "ccache found, so we will use it.")
 endif()
 
+# assume that yardl was run with the following in the _package.yml
+#   sourcesOutputDir: ../cpp/generated/petsird
+
 add_subdirectory(generated/petsird)
+# create a petsird library target
+# Currently it is just empty, but allows creating target properties which are transitive
+add_library(petsird)
+target_link_libraries(petsird PUBLIC petsird_generated)
+target_include_directories(petsird PUBLIC "${PROJECT_SOURCE_DIR}/generated")
+
 add_subdirectory(helpers)

--- a/cpp/README.md
+++ b/cpp/README.md
@@ -17,3 +17,18 @@ The C++ code shows writing to and reading from an HDF5 file
 2. Run the generator: `./petsird_generator test.h5`
 3. Run the analyzer: `./petsird_analysis test.h5`
 4. You can inspect the HDF5 file by running `h5dump test.h5`
+
+## Using this is a library
+
+Currently, we do not install files yet. You therefore need to do something
+like
+```cmake
+set(PETSIRD_dir ../PETSIRD/cpp/) # or wherever
+add_subdirectory(${PETSIRD_dir} PETSIRD)
+
+# only uses petsird
+add_executable(your_executable PUBLIC STIR_PETSIRD_convertor.cpp petsird)
+
+# also uses helpers
+add_library(your_lib PUBLIC petsird_helpers)
+```

--- a/cpp/helpers/CMakeLists.txt
+++ b/cpp/helpers/CMakeLists.txt
@@ -1,12 +1,13 @@
+find_package(xtensor-blas REQUIRED)
+
+# create a petsird_helpers library target
+# Currently it is just empty, but allows creating target properties which are transitive
+add_library(petsird_helpers INTERFACE)
+target_link_libraries(petsird_helpers INTERFACE petsird xtensor-blas)
+target_include_directories(petsird_helpers INTERFACE "${PROJECT_SOURCE_DIR}/helpers/include")
 
 add_executable(petsird_generator petsird_generator.cpp)
-target_link_libraries(petsird_generator petsird_generated)
-target_include_directories(petsird_generator PRIVATE "${PROJECT_SOURCE_DIR}/helpers/include")
-# needed for generated PETSIRD files
-target_include_directories(petsird_generator PRIVATE "${PROJECT_SOURCE_DIR}/generated")
+target_link_libraries(petsird_generator petsird_helpers)
 
 add_executable(petsird_analysis petsird_analysis.cpp)
-target_link_libraries(petsird_analysis petsird_generated)
-target_include_directories(petsird_analysis PRIVATE "${PROJECT_SOURCE_DIR}/helpers/include")
-# needed for generated PETSIRD files
-target_include_directories(petsird_analysis PRIVATE "${PROJECT_SOURCE_DIR}/generated")
+target_link_libraries(petsird_analysis petsird_helpers)

--- a/cpp/helpers/CMakeLists.txt
+++ b/cpp/helpers/CMakeLists.txt
@@ -3,10 +3,10 @@ add_executable(petsird_generator petsird_generator.cpp)
 target_link_libraries(petsird_generator petsird_generated)
 target_include_directories(petsird_generator PRIVATE "${PROJECT_SOURCE_DIR}/helpers/include")
 # needed for generated PETSIRD files
-target_include_directories(petsird_generator PRIVATE "${PROJECT_SOURCE_DIR}/")
+target_include_directories(petsird_generator PRIVATE "${PROJECT_SOURCE_DIR}/generated")
 
 add_executable(petsird_analysis petsird_analysis.cpp)
 target_link_libraries(petsird_analysis petsird_generated)
 target_include_directories(petsird_analysis PRIVATE "${PROJECT_SOURCE_DIR}/helpers/include")
 # needed for generated PETSIRD files
-target_include_directories(petsird_analysis PRIVATE "${PROJECT_SOURCE_DIR}/")
+target_include_directories(petsird_analysis PRIVATE "${PROJECT_SOURCE_DIR}/generated")

--- a/cpp/helpers/include/petsird_helpers.h
+++ b/cpp/helpers/include/petsird_helpers.h
@@ -6,7 +6,7 @@
 #ifndef __petsird_helpers_h__
 #define __petsird_helpers_h__
 
-#include "generated/types.h"
+#include "petsird/types.h"
 #include <array>
 
 namespace petsird_helpers

--- a/cpp/helpers/include/petsird_helpers/create.h
+++ b/cpp/helpers/include/petsird_helpers/create.h
@@ -7,7 +7,7 @@
 #ifndef __petsird_helpers_create_h__
 #define __petsird_helpers_create_h__
 
-#include "generated/types.h"
+#include "petsird/types.h"
 #include <array>
 
 namespace petsird_helpers

--- a/cpp/helpers/include/petsird_helpers/geometry.h
+++ b/cpp/helpers/include/petsird_helpers/geometry.h
@@ -11,7 +11,7 @@
 #include <xtensor/xio.hpp>
 #include <xtensor-blas/xlinalg.hpp>
 #include <vector>
-#include "generated/types.h"
+#include "petsird/types.h"
 #include "petsird_helpers.h"
 
 namespace petsird_helpers

--- a/cpp/helpers/petsird_analysis.cpp
+++ b/cpp/helpers/petsird_analysis.cpp
@@ -9,10 +9,10 @@
 #define USE_HDF5
 
 #ifdef USE_HDF5
-#  include "generated/hdf5/protocols.h"
+#  include "petsird/hdf5/protocols.h"
 using petsird::hdf5::PETSIRDReader;
 #else
-#  include "generated/binary/protocols.h"
+#  include "petsird/binary/protocols.h"
 using petsird::binary::PETSIRDReader;
 #endif
 #include "petsird_helpers.h"

--- a/cpp/helpers/petsird_generator.cpp
+++ b/cpp/helpers/petsird_generator.cpp
@@ -14,10 +14,10 @@
 #define USE_HDF5
 
 #ifdef USE_HDF5
-#  include "generated/hdf5/protocols.h"
+#  include "petsird/hdf5/protocols.h"
 using petsird::hdf5::PETSIRDWriter;
 #else
-#  include "generated/binary/protocols.h"
+#  include "petsird/binary/protocols.h"
 using petsird::binary::PETSIRDWriter;
 #endif
 

--- a/model/_package.yml
+++ b/model/_package.yml
@@ -1,7 +1,7 @@
 namespace: PETSIRD
 
 cpp:
-  sourcesOutputDir: ../cpp/generated
+  sourcesOutputDir: ../cpp/generated/petsird
 
 python:
   outputDir: ../python

